exonum/src/events/codec.rs

```diff
@@ -80,13 +79,30 @@ impl Decoder for MessagesCodec {
             );
         }
 
-        // Read message
-        if buf.len() >= total_len {
-            let data = buf.split_to(total_len).to_vec();
-            let raw = RawMessage::new(MessageBuffer::from_vec(data));
-            return Ok(Some(raw));
+        if total_len > buf.len() {
+            bail!(
+                "Received malicious message with wrong \
+                 total_len: {}, expected message length {}",
+                total_len,
+                buf.len()
+            );
+        }
+
+        let data = buf.split_to(total_len).to_vec();
+        let raw = RawMessage::new(MessageBuffer::from_vec(data));
+        Ok(Some(raw))
+    }
+
+    fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if let Some(frame) = try!(self.decode(buf)) {
+            Ok(Some(frame))
+        } else {
+            if !buf.is_empty() {
+                trace!("Bytes remaining in buffer after receiving EOF.")
```

Comment:
It would be nice to log the number of bytes remaining as it may sometimes give useful insights during debugging. E.g., when there is 1 byte left then one can suspect a trailing null byte or something, when there is a whole megabyte left in the buffer then something is definitely going wrong here.